File: 0012-integer-to-roman.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


char * intToRoman(int num){
    if (num < 1 || num > 3999) return 0;
    int values[] = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
    char *roman_strs[] = {"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};

    char *result = (char *)malloc(16 * sizeof(char));
    memset(result, 0, 16);

    for (int i = 0; i < sizeof(values)/sizeof(int); i++) {
        while (num >= values[i]) {
            num -= values[i];
            strcat(result, roman_strs[i]);
        }
    }

    return result;
}

int main(int argc, char *argv[]) {
    int num = 58;
    printf("%s", intToRoman(num));
    return 0;
}

File: 0012-integer-to-roman.py
class Solution:
    def intToRoman(self, num: int) -> str:
        """
        字符          数值
         I             1
         V             5
         X             10
         L             50
         C             100
         D             500
         M             1000
        """
        roman_identifier = {
            1000: "M",
            900: "CM",
            500: "D",
            400: "CD",
            100: "C",
            90: "XC",
            50: "L",
            40: "XL",
            10: "X",
            9: "IX",
            5: "V",
            4: "IV",
            1: "I",
        }

        result = []
        for k, v in roman_identifier.items():
            if num < k: continue
            k_nums = num // k
            result += [v*k_nums]
            num = num % k

        return "".join(result)
